File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    static class Edge {
        int to;
        int weight;

        Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        List<List<Edge>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] road : roads) {
            int from = road[0];
            int to = road[1];
            graph.get(from).add(new Edge(to, 1));
            graph.get(to).add(new Edge(from, 1));
        }

        int[] answer = new int[sources.length];
        Arrays.fill(answer, -1);

        Map<Integer, Integer> shortestPathFromDestination = new HashMap<>();

        // 목적지로부터 각 지점까지의 최단 거리를 미리 계산
        dijkstra(destination, graph, shortestPathFromDestination);

        // 각 부대원의 출발지에서 목적지까지의 최단 경로를 찾음
        for (int i = 0; i < sources.length; i++) {
            int source = sources[i];
            int shortestPath = shortestPathFromDestination.getOrDefault(source, -1);
            answer[i] = shortestPath;
        }

        return answer;
    }

    // 다익스트라 알고리즘을 이용하여 각 지점으로부터 목적지까지의 최단 거리를 계산
    private void dijkstra(int destination, List<List<Edge>> graph, Map<Integer, Integer> shortestPathFromDestination) {
        PriorityQueue<Edge> pq = new PriorityQueue<>(Comparator.comparingInt(e -> e.weight));
        pq.offer(new Edge(destination, 0));

        while (!pq.isEmpty()) {
            Edge edge = pq.poll();
            int to = edge.to;
            int weight = edge.weight;

            if (shortestPathFromDestination.containsKey(to)) {
                continue; // 이미 최단 거리가 계산된 경우 건너뜀
            }

            shortestPathFromDestination.put(to, weight);

            for (Edge neighbor : graph.get(to)) {
                int nextTo = neighbor.to;
                int nextWeight = weight + neighbor.weight;
                pq.offer(new Edge(nextTo, nextWeight));
            }
        }
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
import heapq

def dijkstra(n, graph, start):
    # 출발 지점부터의 최단 거리를 저장할 리스트
    distances = [float('inf')] * (n + 1)
    distances[start] = 0
    
    # 우선순위 큐를 사용한 Dijkstra 알고리즘
    queue = []
    heapq.heappush(queue, (0, start))
    
    while queue:
        current_distance, current_node = heapq.heappop(queue)
        
        # 현재 노드까지의 거리가 기존에 저장된 거리보다 크다면 스킵
        if current_distance > distances[current_node]:
            continue
        
        # 현재 노드에서 이동 가능한 모든 인접 노드에 대해 최단 거리 갱신
        for adjacent, weight in graph[current_node]:
            distance = current_distance + weight
            if distance < distances[adjacent]:
                distances[adjacent] = distance
                heapq.heappush(queue, (distance, adjacent))
    
    return distances

def solution(n, roads, sources, destination):
    # 인접 리스트 생성
    graph = [[] for _ in range(n + 1)]
    for road in roads:
        a, b = road
        graph[a].append((b, 1))  # 양방향이므로 (b, 1)과 (a, 1) 모두 추가
        graph[b].append((a, 1))
    
    # destination으로부터의 최단 거리 계산
    distances_to_destination = dijkstra(n, graph, destination)
    
    answer = []
    for source in sources:
        # 각 부대원의 위치로부터 destination까지의 최단 거리
        shortest_distance = distances_to_destination[source]
        if shortest_distance == float('inf'):  # 도달할 수 없는 경우
            answer.append(-1)
        else:
            answer.append(shortest_distance)
    
    return answer